### ISSUE TITLE
Flowcharts for Volume Plugin requests

### ISSUE THREAD

**Issue by jschrantz** [docker/docker#21267]
Is there anything similar to the flowcharts on https://docs.docker.com/engine/extend/authorization/ for the volume plugin API?

I am trying to write a simple volume plugin for some internal use cases and it would be nice to see a visual of what requests will be called for the relevant CLI/API commands.

For instance the first time I started testing my plugin I was surprised by the number of Path requests before/after other operations. Based on the documentation I assumed it would have only been used with docker inspect type commands.

**Comment by HackToday** [comment]
> For instance the first time I started testing my plugin I was surprised by the number of Path requests before/after other operations


Could you specify more ? What do you think the issue ?

**Comment by jschrantz** [comment]
I'm not suggesting there is an issue, I just think it would have been useful to know that that is how the docker daemon checks to see if the volume is currently on the machine.

Another unexpected workflow that I came across is that when Mount throws an error, Unmount is called. I had to update my plugin because I assumed the opposite. I'm not suggesting that the behavior is wrong, because it is perfectly fine, but having some documentation for the workflows would have been helpful to me.

**Comment by cpuguy83** [comment]
@jschrantz `Path` is called on `list` (albeit erroneously as we don't use it, and it's a really bad N+1) for `list` calls.
It's also called on container create and volume inspect.
You don't have to return a path if there isn't one yet (ie, it's not mounted).

`Unmount` is called because if there is an issue with `Mount`, docker backs out of the container startup, which may eventually call `Unmount`.

**Comment by jschrantz** [comment]
@cpuguy83 Thanks for the info. By `list` are you referring to `VolumeDriver.List` or is that an alias for some other docker command? If it's the former then that is another one I definitely wouldn't have expected seeing as the `VolumeDriver.List` already contains the Mountpoints for each volume.

Either way it seems that a Flowchart doesn't exist so I'll either have to try digging though the docker code or using a dummy plugin to inspect the calls made for any commands I expect to see.

**Comment by thaJeztah** [comment]
@jschrantz I think such a diagram can be useful; would you be interested in creating it (while you're researching), and opening a pull request? Looks like this website was used to create the diagram you referred to: https://www.websequencediagrams.com. If possible, include the "source" of the diagram (together with the generated image) in your pull-request, so that it's easier to update the image in future

**Comment by jschrantz** [comment]
@thaJeztah I'll try to come up with something.

**Comment by thaJeztah** [comment]
Thanks @jschrantz!

**Comment by suryagaddipati** [comment]
`VolumeDriver.List` -when is this called?. i was expecting to see it being called for `ls` but all I see is  (lots of) `Path` calls .

**Comment by cpuguy83** [comment]
It's called on "ls"

**Comment by suryagaddipati** [comment]
`ls` call still works even after I shut down my driver, i.e it lists the volumes I created when the plugin was running.

Does it store a list somewhere else?

**Comment by cpuguy83** [comment]
Yes it keeps a local cache. Only way to remove the info for now would be to restart the daemon.
1.11 won't cache volumes found from the "ls" call anymore.

**Comment by suryagaddipati** [comment]
oh ok. That explains why `List` was not being called on my plugin ( I am on docker `1.9.1`)

**Comment by suryagaddipati** [comment]
One last question. I see lots of `List` calls. I can't tell why it is being called constantly. I found this in the docs 

`Docker needs reminding of the path to the volume on the host.` 

why does it need `reminding` ?

**Comment by cpuguy83** [comment]
1.9 does not do "list"

**Comment by suryagaddipati** [comment]
Oh sorry I meant `Path` in my previous comment.  Thanks  for your explanation.

**Comment by thaJeztah** [comment]
@jschrantz were you still thinking of working on this?
